cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/ArtifactCacheManager.java

```diff
@@ -0,0 +1,182 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Manages the custom_time and Temporary Holds that need to be set/unset/updated on cached program artifacts in GCS.
+ * Also maintains a usage counter of each of the cached artifacts in the same GCS bucket.
+ */
+public class ArtifactCacheManager {
+
+  private static final Gson GSON = new Gson();
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactCacheManager.class);
+
+  private static final String CACHED_ARTIFACTS_USAGE_COUNT = "cached-artifacts-usage-count.json";
+  private static final String CACHED_ARTIFACTS_LIST_FOR_RUNID = "cached-artifacts-used.json";
+  private static final String CONTENT_TYPE_JSON = "application/json";
+  public static final int MAX_RETRIES_FOR_CACHE_COUNTER_OPERATION = 10;
+  public static final int MAX_RETRIES_TO_FETCH_CACHED_ARTIFACTS_FOR_RUN = 3;
+
+  /**
+   * Increases cache usage counter for all the files(cached artifacts) in the GCS bucket.
+   */
+  public static void recordCacheUsageForArtifacts(Storage client, String bucket, Set<String> files,
+                                                  String runRootPath, String cachePath) {
+    try {
+      String cachedArtifactsListFilePath = getPath(runRootPath, CACHED_ARTIFACTS_LIST_FOR_RUNID);
+      storeCachedArtifactsForRun(client, bucket, files, cachedArtifactsListFilePath);
+      String cacheCountFilePath = getPath(cachePath, CACHED_ARTIFACTS_USAGE_COUNT);
+      changeUsageCountForArtifacts(client, bucket, files, cachePath, cacheCountFilePath, 1);
+    } catch (Exception e) {
+      LOG.error("Error in recordCacheUsageForArtifacts", e);
+    }
+  }
+
+  /**
+   * Decreases cache usage counter for all the files(cached artifacts) in the GCS bucket.
+   * Also updates custom_time and removes Temporary Hold if the artifact is not being used by any program run.
+   */
+  public static void releaseCacheUsageForArtifacts(Storage client, String bucket, String runRootPath) {
+    try {
+      String cachedArtifactsListFilePath = getPath(runRootPath, CACHED_ARTIFACTS_LIST_FOR_RUNID);
+      Set<String> files = getCachedArtifactsForRun(client, bucket, cachedArtifactsListFilePath);
+      String cachedArtifactsPath = getPath(DataprocUtils.CDAP_GCS_ROOT, DataprocUtils.CDAP_CACHED_ARTIFACTS);
+      String cacheCountFilePath = getPath(cachedArtifactsPath, CACHED_ARTIFACTS_USAGE_COUNT);
+      changeUsageCountForArtifacts(client, bucket, files, cachedArtifactsPath, cacheCountFilePath, -1);
+    } catch (Exception e) {
+      LOG.error("Error in releaseCacheUsageForArtifacts", e);
+    }
+  }
+
+  private static void changeUsageCountForArtifacts(Storage client, String bucket, Set<String> files, String cachePath,
```

Comment:
nit: this method is very long. Is it possible to extract certain logics and put them in smaller methods?